app/gui/src/controller/searcher/component/hardcoded.rs

```diff
@@ -34,30 +26,92 @@ thread_local! {
-            .with_return_type("Standard.Base.Data.Text.Text")
+        Snippet::new_with_name_and_code_and_icon("text input", "\"\"", IconId::TextInput)
+            .with_return_types(["Standard.Base.Data.Text.Text"])
             .with_documentation(
                 "A text input node.\n\n\
-                An empty text. The value can be edited and used as an input for other nodes."
+                An empty text. The value can be edited and used as an input for other nodes.",
             )
             .into(),
-        snippet_with_name_and_code_and_icon("number input", "0", IconId::NumberInput)
-            .with_return_type("Standard.Base.Data.Numbers.Number")
+        Snippet::new_with_name_and_code_and_icon("number input", "0", IconId::NumberInput)
+            .with_return_types([
+                "Standard.Base.Data.Numbers.Number",
+                "Standard.Base.Data.Numbers.Decimal",
+                "Standard.Base.Data.Numbers.Integer",
+            ])
             .with_documentation(
-                 "A number input node.\n\n\
-                 A zero number. The value can be edited and used as an input for other nodes."
+                "A number input node.\n\n\
+                 A zero number. The value can be edited and used as an input for other nodes.",
             )
             .into(),
     ];
 }
 
 
-// === Constants helpers ===
+// === Constants utilities ===
+
+/// Return a filtered copy of [`INPUT_SNIPPETS`] containing only snippets which have at least one
+/// of their return types on the given list of return types.
+pub fn input_snippets_with_matching_return_type(
+    return_types: impl IntoIterator<Item = tp::QualifiedName>,
+) -> Vec<Rc<Snippet>> {
+    let rt_set: HashSet<_> = return_types.into_iter().collect();
+    let rt_of_snippet_is_in_set =
+        |s: &&Rc<Snippet>| s.return_types.iter().any(|rt| rt_set.contains(rt));
+    INPUT_SNIPPETS
+        .with(|snippets| snippets.iter().filter(rt_of_snippet_is_in_set).cloned().collect_vec())
+}
+
+
+
+// ===============
+// === Snippet ===
+// ===============
+
+/// A hardcoded snippet of code with a description and syntactic metadata.
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct Snippet {
+    /// The name displayed in the [Component Browser](crate::controller::searcher).
+    pub name:               &'static str,
+    /// The code inserted when picking the snippet.
+    pub code:               &'static str,
+    /// The type returned by the snippet's code.
+    pub return_types:       Vec<tp::QualifiedName>,
+    /// The documentation bound to the snippet.
+    pub documentation_html: Option<String>,
+    /// The ID of the icon bound to this snippet's entry in the [Component
+    /// Browser](crate::controller::searcher).
+    pub icon:               IconId,
+}
+
+impl Snippet {
+    /// Construct a hardcoded snippet with given name, code, and icon.
+    fn new_with_name_and_code_and_icon(
```

Comment:
Now I think if we could call it just "new", as we have no other constructor variants?